stan/math/prim/prob/multi_normal_cholesky_lpdf.hpp

```diff
@@ -119,59 +122,49 @@ return_type_t<T_y, T_loc, T_covar> multi_normal_cholesky_lpdf(
   }
 
   if (include_summand<propto, T_y, T_loc, T_covar_elem>::value) {
-    Eigen::Matrix<T_partials_return, Eigen::Dynamic, Eigen::Dynamic>
-        y_val_minus_mu_val(size_y, size_vec);
+    row_vector_partials_t half(size_vec);
+    vector_partials_t y_val_minus_mu_val(size_vec);
+    vector_partials_t scaled_diff(size_vec);
+    matrix_partials_t L_val = value_of(L_ref);
+
+    T_partials_return sum_lp_vec(0.0);
 
     for (size_t i = 0; i < size_vec; i++) {
       decltype(auto) y_val = as_value_column_vector_or_scalar(y_vec[i]);
       decltype(auto) mu_val = as_value_column_vector_or_scalar(mu_vec[i]);
-      y_val_minus_mu_val.col(i) = y_val - mu_val;
+      y_val_minus_mu_val = eval(y_val - mu_val);
+      half = mdivide_left_tri<Eigen::Lower>(L_val, y_val_minus_mu_val)
+                 .transpose();
+      scaled_diff = mdivide_right_tri<Eigen::Lower>(half, L_val).transpose();
```

Comment:
This is the part that concerns me, since it's gone from a single solve (with a matrix) to `size_vec` solves (with a vector) for `half` and `scaled_diff` each. Especially when the single larger solve can be better vectorised with SIMD & other compiler opts

Is there enough of a memory hit to justify to extra operations?

Reply:
Definitely agreed. We do this sort of thing already in
* https://github.com/stan-dev/math/blob/0c16b89a2bbf0153ebf8d435c416cd626506558d/stan/math/prim/prob/multi_normal_lpdf.hpp#L94-L98
* https://github.com/stan-dev/math/blob/0c16b89a2bbf0153ebf8d435c416cd626506558d/stan/math/prim/prob/multi_student_t_lpdf.hpp#L124-L128

I'm really not sure what's best here.

Reply:
If it's alright with you, I'd prefer not to implement this. The current implementation is likely to scale better to larger inputs, and the changes would also reduce any benefits from OpenCL accelerated-ops.

But also completely happy for you to call someone in for a tie-breaker if you feel strongly about it!

Reply:
I'm fine with closing it but I want someone to weigh in on if we should change the other distributions. I can update the mvn derivatives pr to follow the same.

Reply:
@SteveBronder - as the Chief of Memory Police, what do you think?

Reply:
imo if the benchmarks are not showing a huge difference I think it's good. 

The main thing I'm thinking of is memory pressure. While this code is going to produce more instructions that need to be run for each loop, the memory it's working over will be smaller and more likely to stay in cache. 

@andrjohns I think the memory pressure is more of a focus because either way this is implemented, if a machine has large simd instructions the vector code should still use them. 

@spinkney can you run the benchmark comparing two `var` inputs (and `var<Matrix>` inputs)  like 30 or so times and just plot that? imo I think that's what we care about the most. I have a little repo below you can use to compare both branches. Also let us know what cpu you are using. I can test this on my m2 mac as well if you send me a script to do it.

https://github.com/SteveBronder/stan-perf

Reply:
I have an M1max. Is there someone who could benchmark on a windows and linux machine?

Reply:
I got the library setup but I don't have taskset. Also, how can I set up the script to run the two branches?

Reply:
You don't need taskset to run the benchmarks, only if you want to isolate in a single core. 

You can add another branch in your benchmarks cmake file like 

```
FetchContent_Declare(
  stanmathalt
  GIT_REPOSITORY https://github.com/stan-dev/math
  GIT_TAG  mybranch # replace with the version you want to use
)

FetchContent_GetProperties(stanmathalt)
if(NOT stanmathalt_POPULATED)
  FetchContent_Populate(stanmathalt)
endif()

```

Then you can include it in your executible build like `${stanmathalt_SOURCE_DIR}`

Reply:
Then how do I just run benchmarks for this distribution?

Reply:
There's an example of how to add a benchmark on the readme and an example benchmark folder below. You need to write a little Cmake file for compiling the benchmark and should be able to use that folder as an example 

https://github.com/SteveBronder/stan-perf/tree/main/benchmarks/matmul_aos_soa